File: merga-2d-arr-sum/main.go
package main

import (
	"sort"
)

/**
 * source: https://leetcode.com/problems/merge-two-2d-arrays-by-summing-values/
 * problem: https://leetcode.com/problems/merge-two-2d-arrays-by-summing-values/
 * type: Easy
 * site: LeetCode
 * submission: https://leetcode.com/problems/merge-two-2d-arrays-by-summing-values/submissions/1642477815/
 */

func mergeArrays(nums1 [][]int, nums2 [][]int) [][]int {
	counter := make(map[int]int)
	ans := [][]int{}

	for idx := range nums1 {
		key, val := nums1[idx][0], nums1[idx][1]

		counter[key] = val
	}

	for idx := range nums2 {
		key, val := nums2[idx][0], nums2[idx][1]

		counter[key] += val
	}

	for key, val := range counter {
		temp := []int{key, val}
		ans = append(ans, temp)
	}

	sort.Slice(ans, func(i, j int) bool {
		return ans[i][0] < ans[j][0]
	})

	return ans
}

func main() {
	nums1 := [][]int{{1, 2}, {2, 3}, {4, 5}}
	nums2 := [][]int{{1, 4}, {3, 2}, {4, 1}}
	println(mergeArrays(nums1, nums2))

	nums1 = [][]int{{2, 4}, {3, 6}, {5, 5}}
	nums2 = [][]int{{1, 3}, {4, 3}}
	println(mergeArrays(nums1, nums2))

}

File: merga-2d-arr-sum/main.py
#!/usr/bin/env python3

"""
source: https://leetcode.com/problems/merge-two-2d-arrays-by-summing-values/
problem: https://leetcode.com/problems/merge-two-2d-arrays-by-summing-values/
type: Easy
site: LeetCode
submission: https://leetcode.com/problems/merge-two-2d-arrays-by-summing-values/submissions/1642465230/
"""


class Solution:
    # def mergeArrays(self, nums1: List[List[int]], nums2: List[List[int]]) -> List[List[int]]:
    def mergeArrays(self, nums1, nums2):
        ans = {x[0]: x[1] for x in nums1}
        for x in nums2:
            key, val = x
            ans[key] = ans.get(key, 0)+val

        return sorted([[key, val] for key, val in ans.items()])


def main():
    sol = Solution()

    nums1 = [[1, 2], [2, 3], [4, 5]]
    nums2 = [[1, 4], [3, 2], [4, 1]]
    print(sol.mergeArrays(nums1, nums2))

    nums1 = [[2, 4], [3, 6], [5, 5]]
    nums2 = [[1, 3], [4, 3]]
    print(sol.mergeArrays(nums1, nums2))

    pass


if __name__ == "__main__":
    main()
